src/parquet/bloom_filter.cc

```diff
@@ -94,30 +71,52 @@ BloomFilter::BloomFilter(const uint32_t *bitset, uint32_t num_bytes)
       throw parquet::ParquetException("Unsupported hash strategy");
   }
 
+  if (!bitset) {
+    throw parquet::ParquetException("bitset pointer is NULL");
+  }
+
+  if (num_bytes < MINIMUM_BLOOM_FILTER_BYTES || num_bytes > MAXIMUM_BLOOM_FILTER_BYTES ||
+      (num_bytes & (num_bytes - 1)) != 0) {
+    throw parquet::ParquetException("Given length of bitset is illegal");
+  }
+
   switch (algorithm_) {
     case Algorithm::BLOCK:
-      this->bloom_algorithm_.reset(new BlockBasedAlgorithm());
+      this->bloom_filter_algorithm_.reset(new BlockBasedAlgorithm(bitset, num_bytes));
       break;
     default:
       throw parquet::ParquetException("Unsupported Bloom filter algorithm");
   }
+}
 
-  if (!bitset) {
-    throw parquet::ParquetException("bitset pointer is NULL");
+BloomFilter* BloomFilter::Deserialize(InputStream* input) {
+  int64_t bytes_available;
+
+  uint32_t len;
+  memcpy(&len, input->Read(sizeof(uint32_t), &bytes_available), sizeof(uint32_t));
```

Comment:
Can `Read` return a `nullptr`? If so, you'll want to checkthat the result is non-null before calling `memcpy`.